### PR TITLE
[TypeChecker] Implement result builder transform via AST modification

### PR DESCRIPTION
**Pull request by xedin**
The transform changes closure body into a multi-statement closure
with all of the implicit result builder calls and type-checks it
like a regular closure. After transformation of a particular builder to a
particular function body is applied, it is cached and reused in case
other overloads require the same transform.

There are a couple of result builder specific changes mentioned below,
otherwise the logic to generate constraints and apply solutions is
unchanged:

- Placeholder variable: A variable declaration that doesn't have a
   type and infers it from first use site. If such a variable
   has an initializer, it would be type-checked during solution application.

 - `TypeJoinExpr` - an implicit expression that refers to a "join" variable
    and a set of expressions that should all produce the same type that
    becomes a type of a "join" variable.

Diagnostics: Since result builders are type-checked as regular multi-statement
closures it means that first error in the body skips the rest of it, which helps to
pinpoint issues and improve performance of an invalid code.

Resolves: rdar://85265414

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Comment by xedin** [comment]
@swift-ci please test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
Reduced example from the issue found by the stress tester:

```swift
// RUN: %swift-ide-test -code-completion -source-filename %s -code-completion-token COMPLETE

protocol View {}

@resultBuilder
struct ViewBuilder {
  static func buildBlock(_ component: MyView) -> MyView {
    MyView()
  }
}

struct QStack<Content> {
  init(@ViewBuilder content: () -> Content) {}
  
  func qheet<Content>(@ViewBuilder content: @escaping () -> Content) -> some View {
    MyView()
  }
}

struct MyView : View {
  func qag<V>(_ tag: V) -> MyView {
    MyView()
  }
}


func foo() {
  QStack {
    MyView().qag(0)
  }
  .qheet() {
    MyView().#^COMPLETE^#
  }
}
```

used to produce a result `qag` but doesn’t produce any results with this PR.

**Comment by xedin** [comment]
@ahoppen Thank you! I'll look into that

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -5121,6 +5121,11 @@ namespace {
       return E;
     }
 
+    Expr *visitTypeJoinExpr(TypeJoinExpr *E) {
+      // Type-join expressions should not appear in the type-checked AST.
```
Comment (LucianoPAlmeida) [inline lib/Sema/CSApply.cpp:5125]: It is possible have a `TypeJoinExpr` here due to a bug? If it is maybe is worth assert or `llvm_unreachable` so it is noticeable?

File: lib/Sema/CSStep.cpp [hunk 2]
```diff
@@ -958,24 +955,47 @@ StepResult ConjunctionStep::resume(bool prevFailed) {
           // of the constraint system, so they have to be
           // restored right afterwards because score of the
           // element does contribute to the overall score.
-          restoreOriginalScores();
+          restoreBestScore();
+          restoreCurrentScore(solution.getFixedScore());
+
+          // Transform all of the unbound outer variables into
+          // placeholders since we are not going to solve for
+          // each ambguous solution.
```
Comment (LucianoPAlmeida) [inline lib/Sema/CSStep.cpp:963]:
```suggestion
          // each ambiguous solution.
```

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSApply.cpp:5125]: Yes, thanks for spotting this! Not having llvm_unreachable here is left-over from initial changes that introduced TypeJoinExpr

File: lib/Sema/BuilderTransform.cpp [hunk 3]
```diff
@@ -723,178 +1475,167 @@ class BuilderClosureVisitor
...
+  UNSUPPORTED_STMT(DoCatch)
+  UNSUPPORTED_STMT(RepeatWhile)
+  UNSUPPORTED_STMT(Break)
+  UNSUPPORTED_STMT(Continue)
+  UNSUPPORTED_STMT(Fallthrough)
+  UNSUPPORTED_STMT(Fail)
+  UNSUPPORTED_STMT(PoundAssert)
+  UNSUPPORTED_STMT(Case)
 
 #undef CONTROL_FLOW_STMT
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/BuilderTransform.cpp:1640]: It should actually be only `#undef UNSUPPORTED_STMT` here

[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/BuilderTransform.cpp]: Fixed, thank you!

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSApply.cpp]: Fixed!

File: lib/Sema/CSStep.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSStep.cpp]: Fixed!

**Comment by xedin** [comment]
@ahoppen The reason why your example would no longer produce results is related to filtering - in code completion mode the filtering is disabled so first builder would end up with two solutions for `0` - `Int` and `Double` and that means that `MyView().qag(0)` marked as ambiguous and solver just skips the rest of the constraint system. I think the way to fix this would be to allow filtering everywhere in the builder body besides syntatic element that has code completion token, WDT?

**Comment by ahoppen** [comment]
I see. I think your suggestion makes perfect sense and I already played around with it in https://github.com/apple/swift/pull/42024. I would be fine with merging the PR as-is and taking care of that problem later on.

**Comment by xedin** [comment]
Sounds good!

**Comment by xedin** [comment]
@swift-ci please build toolchain

**Comment by xedin** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by xedin** [comment]
@swift-ci please toolchain macosx

**Comment by xedin** [comment]
@swift-ci please build toolchain macos

**Comment by xedin** [comment]
@ahoppen Looks like changes to add source range to an implicit variable are causing verifier failures.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by xedin** [comment]
@swift-ci please smoke test macOS

**Comment by xedin** [comment]
@swift-ci Please SourceKit stress test

**Comment by xedin** [comment]
Closing this one since we are going to land it piecemeal starting from https://github.com/apple/swift/pull/60065

**Pull request closed without merging**